flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/FinishedOperatorChain.java

```diff
@@ -99,5 +101,29 @@ public void snapshotState(
             Supplier<Boolean> isRunning,
             ChannelStateWriter.ChannelStateWriteResult channelStateWriteResult,
             CheckpointStreamFactory storage)
-            throws Exception {}
+            throws Exception {
+        for (StreamOperatorWrapper<?, ?> operatorWrapper : getAllOperators(true)) {
+            StreamOperator<?> operator = operatorWrapper.getStreamOperator();
+
+            if (operator == getMainOperator() || operator == getTailOperator()) {
+                OperatorSnapshotFutures snapshotInProgress = new OperatorSnapshotFutures();
+                if (operator == getMainOperator()) {
+                    snapshotInProgress.setInputChannelStateFuture(
+                            channelStateWriteResult
+                                    .getInputChannelStateHandles()
+                                    .thenApply(StateObjectCollection::new)
+                                    .thenApply(SnapshotResult::of));
+                }
+                if (operator == getTailOperator()) {
+                    snapshotInProgress.setResultSubpartitionStateFuture(
+                            channelStateWriteResult
+                                    .getResultSubpartitionStateHandles()
+                                    .thenApply(StateObjectCollection::new)
+                                    .thenApply(SnapshotResult::of));
+                }
```

Comment:
Could you refactor the code and deduplicate this with `RegularOperatorChain#buildOperatorSnapshotFutures`?

For example this highlighted code could be extracted to a helper method in `OperatorChain` and used both in the `Regular` and `Finished` versions?